src/lib/merkle_ledger/database.ml

```diff
@@ -684,16 +694,31 @@ module Make (Inputs : Inputs_intf) :
       else location
     in
     assert (Location.is_hash location) ;
-    let rec loop k =
-      if Location.height ~ledger_depth:mdb.depth k >= mdb.depth then []
-      else
-        let sibling = Location.sibling k in
-        let sibling_dir = Location.last_direction (Location.to_path_exn k) in
-        let hash = get_hash mdb sibling in
-        Direction.map sibling_dir ~left:(`Left hash) ~right:(`Right hash)
-        :: loop (Location.parent k)
+    let rev_locations, rev_directions =
+      let rec loop k loc_acc dir_acc =
+        if Location.height ~ledger_depth:mdb.depth k >= mdb.depth then
+          (loc_acc, dir_acc)
+        else
+          let sibling = Location.sibling k in
+          let sibling_dir = Location.last_direction (Location.to_path_exn k) in
+          loop (Location.parent k) (sibling :: loc_acc) (sibling_dir :: dir_acc)
+      in
+      loop location [] []
+    in
+    let rev_hashes = get_hash_batch mdb rev_locations in
+    let rec loop directions hashes acc =
+      match (directions, hashes) with
+      | [], [] ->
+          acc
+      | direction :: directions, hash :: hashes ->
+          let dir =
+            Direction.map direction ~left:(`Left hash) ~right:(`Right hash)
+          in
+          loop directions hashes (dir :: acc)
+      | _ ->
+          failwith "Mismatched lengths"
```

Comment:
Can we use `map` and `unzip` here instead? I know the performance is slightly worse but I think we should optimize for clarity over smaller performance gains.